### PR TITLE
Image: Avoid link class loss when pasting for raw transformation

### LINKED ISSUE #11995: Raw transforms of linked image should support custom link classes
**Is your feature request related to a problem? Please describe.**

Now that image links can have a custom class assigned (see #7771), we should support existing linked images' link classes.

**Describe the solution you'd like**

Proposed solution identified by @talldan in https://github.com/WordPress/gutenberg/pull/7771#discussion_r231757707.

**Describe alternatives you've considered**

I have not explored alternatives yet.

### PR DESCRIPTION
**Pull request by Mamaduka**
## What?
Fixes #11995.

PR updates the Image block's raw transformation schema to support link classes. This prevents content loss when pasting an image.

Users can edit these classes using Link UI. See #7771.

## Testing Instructions
1. Create a post or page.
2. Add an empty paragraph.
3. Paste test image content.

<details><summary>Test image HTML</summary>
<p>

```html
<figure class="wp-block-image size-large"><a class="test-link hey --what" href="https://bbc.com"><img src="https://s.w.org/patterns/files/2021/06/image-from-rawpixel-id-539759-jpeg-1-1024x1024.jpg" alt="Vintage Cupid Illustration" class="wp-image-179"/></a></figure>
``` 

</p>
</details> 

### Testing Instructions for Keyboard
Same.

## Screenshot
CleanShot 2024-12-10 at 22 15 04

### CODE REVIEW

File: packages/block-library/src/image/transforms.js [hunk 1]
```diff
@@ -59,6 +59,7 @@ const schema = ( { phrasingContentSchema } ) => ( {
 			...imageSchema,
 			a: {
 				attributes: [ 'href', 'rel', 'target' ],
+				classes: [ /[\w-]*/ ],
```
Comment (Mamaduka) [inline packages/block-library/src/image/transforms.js:62]: Maybe we should support a wildcard matcher in `cleanNodeList` for `class`. There's already support for it for `children` property.

https://github.com/WordPress/gutenberg/blob/467e4a22b6875aef32c6baf09ef38f62eeba2edd/packages/dom/src/dom/clean-node-list.js#L109-L112

Comment (Mamaduka) [reply packages/block-library/src/image/transforms.js:62]: Here's the follow-up #67830.

**Review by t-hamano — COMMENTED** [review]
Would it be better to close this PR and proceed with #67830? Or would it be better to proceed with this PR first and then proceed with #67830?

**Comment by Mamaduka** [comment]
The #67830 is just a proposal; not sure if the wildcard support is needed there yet. Only created it to open the conversation.

**Review by t-hamano — APPROVED** [review]
LGTM! Japanese characters and emojis also seem to work correctly:

```html
<figure class="wp-block-image size-large">
	<a class="test-link hey --what test_link &<> あああ 😄" href="https://bbc.com">
		<img src="https://s.w.org/patterns/files/2021/06/image-from-rawpixel-id-539759-jpeg-1-1024x1024.jpg" alt="Vintage Cupid Illustration" class="wp-image-179"/>
	</a>
</figure>
```

Nit: Removing the hyphen from the regex pattern appears to be fine.

```diff
diff --git a/packages/block-library/src/image/transforms.js b/packages/block-library/src/image/transforms.js
index 32824c372e..207f647b4a 100644
--- a/packages/block-library/src/image/transforms.js
+++ b/packages/block-library/src/image/transforms.js
@@ -59,7 +59,7 @@ const schema = ( { phrasingContentSchema } ) => ( {
                        ...imageSchema,
                        a: {
                                attributes: [ 'href', 'rel', 'target' ],
-                               classes: [ /[\w-]*/ ],
+                               classes: [ /[\w]*/ ],
                                children: imageSchema,
                        },
                        figcaption: {
```

**Comment by Mamaduka** [comment]
> Removing the hyphen from the regex pattern appears to be fine.

In my tests, it breaks hyphenated classes. You can do a quick Regex test on https://regex101.com.

**Pull request merged**